### PR TITLE
Refactor CDO application config

### PR DESCRIPTION
**Pull request by wjordan**
This PR refactors the global `CDO` application configuration currently contained in `deployment.rb` into environment-specific, ERB+YAML-based files more consistent with `locals.yml` / `globals.yml` configuration format:

- `config.yml.erb` - global application config
- `config/[env].yml.erb` - environment-specific application config

The new config-loading order is defined in `cdo.rb` (1 = highest priority):
1. `ENV` - environment variables (`CDO_*`, plus `RAILS_ENV` / `RACK_ENV`)
2. `locals.yml` - local configuration
3. `globals.yml` - [Chef-]provisioned configuration
4. `config/[env].yml.erb` - environment-specific defaults
5. `config.yml.erb` - global defaults

This PR also makes the loader more strict and reliable, by enforcing that `CDO` variables can't be set from anywhere else in the application after the configuration has been loaded. All variables must be declared in `config.yml.erb`, so we will be able to better maintain a reliable, (increasingly well-documented/commented) list of all configuration variables existing throughout our application.

This also required some changes to various tests that stub certain CDO variables during their test.

### CODE REVIEW

**Review by uponthesun — COMMENTED** [review]
Have not looked through the code changes yet, but I think it's very important to include a "developer's manual" documentation file with this change. This can include:

* syntax
* explanation of interaction with chef / AWS secrets
* best practices for what should / shouldn't live in this config

File: lib/cdo/config.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    # but allow stubbing for unit tests.
+    def freeze
+      @table.each_key(&method(:new_ostruct_member!))
+      @frozen = true
+    end
+
+    def method_missing(key, *args)
+      msg = "Undefined #{self.class} reference: #{key}"
+      raise ArgumentError, msg, caller(1) if @frozen
+      # puts msg, caller(1) if @frozen
```
Comment (uponthesun) [inline lib/cdo/config.rb]: Remove? (And in `modifiable?` also)

File: config/test.yml.erb [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# Unit tests need to stub a lot of CDO properties so don't freeze config.
```
Comment (uponthesun) [inline config/test.yml.erb]: What part of the code/config makes it not freeze config for test?

File: lib/test/cdo/test_config.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+require_relative '../test_helper'
+require 'cdo/config'
+require 'tempfile'
+
+class ConfigTest < Minitest::Test
```
Comment (uponthesun) [inline lib/test/cdo/test_config.rb:5]: Some test cases I'd like to see:

* Resolving transitive dependencies successfully (like `test_circular_dependency` but without the circular-ness)
* `freeze` functionality

**Review by uponthesun — COMMENTED** [review]
Looks good to me overall code-wise, just a few minor comments here. As I mentioned in slack, I think we should also have an idea of our confidence level in the config translation and assess potential risks before deploying.

File: config/test.yml.erb (hunk 2 above)
Comment (wjordan) [reply config/test.yml.erb]: oh this was a leftover comment (I had toyed with the idea of making `freeze` configurable, then decided to always freeze for simplicity), thanks for pointing it out, I will remove.

File: lib/test/cdo/test_config.rb (hunk 3 above)
Comment (wjordan) [reply lib/test/cdo/test_config.rb:5]: - `test_erb` includes a transitive (`c -> b -> a`) dependency test along those lines
- I will add a test for `#freeze`

File: lib/cdo/config.rb (hunk 1 above)
Comment (wjordan) [reply lib/cdo/config.rb]: oh yeah, that was leftover from manual testing, I will clean this up.

**Comment by wjordan** [comment]
For extra confidence (beyond the passing CI test build) that this refactoring PR doesn't unintentionally change any existing config behavior, I ran the following manual config-comparison test to compare the default configurations loaded after this PR match the baseline configurations:

```
export RAILS_ENV=test
git checkout staging; bundle exec ruby -r./deployment -e "puts CDO.to_h.compact.sort.to_h.to_json" | jq . > /tmp/staging.json
git checkout cdo-config-2; bundle exec ruby -r./deployment -e "puts CDO.to_h.compact.sort.to_h.to_json" | jq . > /tmp/cdo-config.json
diff /tmp/staging.json /tmp/cdo-config.json
```
(repeat with `export RAILS_ENV=development`, `production`, `staging`)

The only diffs this outputs are expected changes from added extra defaults (e.g., `curriculum_languages`, various `chef_*` keys), and one removed unused key (`languages`).

**Review by uponthesun — APPROVED** [review]
Looks great!

**Pull request merged**